Added CoSEConstants.TILING_COMPARE_BY

metincansiper opened the pull request:
Works with https://github.com/cytoscape/cytoscape.js-cose-bilkent/pull/116

hasanbalci commented:
Hi @metincansiper, I liked this option since it will give more flexibility to users to sort the nodes to be tiled.  I will review the changes in detail, but can you please make them in unstable branches?

metincansiper commented:
Hi @hasanbalci, thanks! I realized that there would be one detail that I would better revise and update if needed. I will make it and make the PR's to unstable branch after that.

metincansiper commented:
@hasanbalci I updated the base branches of this PR and the one related to cytoscape.js-cose-bilkent repository (https://github.com/cytoscape/cytoscape.js-cose-bilkent/pull/116) as ``unstable``.
 
However, I realized something in this repository. It happens both the branch in my fork after making it based on unstable and the original unstable branch of this repository. I will report the output from the unstable branch of this repository:

When I use the node version 13.6.0 at the build step (``npm run build``) I see the following error: ``(node:34424) UnhandledPromiseRejectionWarning: Error: No valid exports main found for '/Users/siperm/Documents/Workspace/cose-base/node_modules/colorette'``.

Then I updated my node version as 15.13.0 and the build was successful but I see that the bundled ``cose-base.js`` file is produced as minified. However, the existing ``cose-base.js`` file is not minified in the unstable branch.

Then I tried the node version 8.2.1 and at the build step I see the error of ``SyntaxError: Unexpected token ...``.

Is there a specific version of node that must be used?

hasanbalci commented:
@metincansiper It seems that we updated webpack version, but didn't configure webpack.config.js accordingly. I use node version 13.10.1 and it didn't give error but minified the `cose-base.js` file. Thanks for reporting this. I will make the required changes after I reviewed and merged your PR.

hasanbalci commented on src/CoSELayout.js:
```diff
@@ -1439,12 +1560,23 @@ CoSELayout.prototype.tileNodes = function (nodes, minWidth) {
+  };
+
   // Sort the nodes in ascending order of their areas
   nodes.sort(function (n1, n2) {
```
Here, it would be better to keep sorting in descending order. Especially sorting in descending order when the node area is used (default) generally gives better results. The comment from the past might be written wrong. We may also add an explanation for `tilingSortBy` option in extension side to clarify that it sorts in descending order based on the returned value.

metincansiper replied:
@hasanbalci then I think there are two options. Either of them works for me. I can update it as whichever you prefer:

1) As you mentioned restore default option (so that it sorts in descending order of areas) and also make ``tilingSortBy`` option working similarly meaning that sorting in descending order by default and updating the explanation accodingly.
2) Restore the default option (again so that it sorts in descending order of areas) but make the optional ``tilingSortBy`` option as it currently is so that it would sort in ascending order of return values of ``tilingSortBy`` function.

hasanbalci replied:
Let's apply the first one to be consistent between two options. If user wants to sort in ascending order, he/she can do that easily by multiplying the value with -1 before returning it.

metincansiper replied:
@hasanbalci I guess with this option there would be a lot of cases where the return value of the function will not be a number. If it is a string I think there may be a need for reversing the string instead of multiplying by -1. I actually do not know if it would be an issue for some users as the strings get longer and number of nodes increase. 

If it sounds okay I would add an extra option like ``tilingSortByOrder`` which is ``descending`` by default. Node areas will be sorted in descending order anyways this option is only to work together with ``tilingSortBy`` option anyways.

hasanbalci replied:
@metincansiper Ok, then let's do it as you suggested.

metincansiper replied:
@hasanbalci maybe before moving on I can make another suggestion about this part. Maybe we can consider having another option like ``tilingCompareBy`` instead of these two that I mentioned (``tilingSortBy`` and  ``tilingSortByOrder``). 

It may work similar to optional ``compareFunction`` of javascript's ``Array.prototype.sort()`` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort).

I think that advantage of this option over the other two options we discussed maybe enabling users to have more customizable sorting. For example when users are working with mix of strings and numbers javascript's grater than operator will return true for ``"a2" > "a11"``. However, I suspect some users may prefer "a11" considered bigger than "a2" for sorting. Also, maybe having just one option would be simpler to manage.

What do you think?

hasanbalci commented on src/CoSELayout.js:
```diff
@@ -1425,8 +1425,129 @@ CoSELayout.prototype.tileCompoundMembers = function (childGraphMap, idToNode) {
+    // if horizontalCount count is not a float value then both of rounding to floor and ceil
+    // will yield the same values. Instead of repeating the same calculation try going up
+    // while favoring horizontal dimension in such cases
+    if ( horizontalCount == horizontalCountDouble ) {
```
Here, let's say `horizontalCountDouble` is equal to 4. Then `horizontalCount` will also be 4. Since they are equal, we make `horizontalCount` 5 in below line. Is this what we want or do we want to keep it as 4 since it's not a float?

metincansiper replied:
@hasanbalci actually we calculate 2 values one by favoring horizontal dimension and the other one by favoring vertical dimension. The base call are made around here (https://github.com/iVis-at-Bilkent/cose-base/blob/89fdaad9669c9fe7c33f800726070cd91e33a88f/src/CoSELayout.js#L1428). 

What we do around the lines that you commented is the way we are doing that favoring the specified dimension in the method. When ``horizontalCountDouble`` is a float value like 4.1 if the dimension to favor is "horizontal" we round the value to ceil, when it is "vertical" we round it to floor. In that way we compare the results where the "horizontalCount" is 4 and 5. Then we choose the one that yields to the best aspect ratio.

Let's say the ``horizontalCountDouble`` is not a float value but it is an integer value like 4. We still use the one up which is 5 when we are favoring horizontal side and we are using the value as is which is 4 when we are favoring the vertical dimension. And then again we are choosing the one yielding the best aspect ratio.

In this way we are increasing the chance of having a more desired aspect ratio.

hasanbalci replied:
I got it, no problem.

hasanbalci reviewed (commented):
Hi @metincansiper, it seems ok. I just added two comments, one for clarifying `horizontalCount` calculation and the other one is for changing sorting order. Thanks!

metincansiper commented:
@hasanbalci I added my responses/questions below your comments.

hasanbalci commented:
@metincansiper I can't see your responses.

metincansiper commented:
@hasanbalci sorry I forgot to submit it. I think it must be visible now.

metincansiper commented:
@hasanbalci I made the updates that we have discussed. BTW as we discussed before the auto generated ``cose-base.js `` file ended up being minimized in this commit.

hasanbalci commented on src/CoSELayout.js:
```diff
@@ -1439,13 +1560,26 @@ CoSELayout.prototype.tileNodes = function (nodes, minWidth) {
+      cmpBy = tilingCompareBy;
+    }
+
+    return cmpBy( n1, n2 );
```
@metincansiper I think we need to get `n1.id` and `n2.id` as parameters of `cmpBy` function if `organization.idealRowWidth`exists, right? Because, we expect users to use node ids as the parameter of `tilingCompareBy` function.

metincansiper replied:
@hasanbalci thanks for noticing it. I fixed it. Looks like I missed it since in my test file nodes were sorted in ascending order even with this bug. Along with that I found another issue in the test/demo file and fixed it as well.
